perf: fetching home page

flagrede opened the pull request:
## Description

Closes: regen-network/regen-registry#1313

- improve data fetching for the home page, the following requests have been cached
  - sanity AllHomePage
  - sanity AllCreditClasses
  - ledger projects
  - ledger sellOrders
  - ledger allowedDenom 
  - ledger metadata
- cache additional queries on credit batches page
  - ledger project
  - ledger metadata 
- add home loader to trigger fetch early
- type queries for react-query properly
- simplify queries creation 
- use only one instance of ReactQuery across the app
- separate logic from fetching for home page
  - have a look at [this file](https://github.com/regen-network/regen-web/blob/03ecd45048530d7c61739ea9c08da2a545d8e21e/web-registry/src/hooks/projects/useProjectsWithOrders.ts) regarding that  
- add react-query devtool (only imported in dev env)
![image](https://user-images.githubusercontent.com/530644/203818744-b9a85537-82de-4b75-bd79-71336071918d.png)


---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] provided a link to the relevant issue or specification
- [x] provided instructions on how to test
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed
- [ ] once the PR is closed, set up backport PRs for `redwood`, `hambach` and optionally `master` (see below)

#### Setting up backport PRs

After merging your PR to `dev`, set up backports by doing the following:

1. If your branch does not have merge commits, add the following comment to
   your PR:

- If PR is a new feature: `@Mergifyio backport redwood hambach`.
- If PR is a bug fix: `@Mergifyio backport redwood hambach master`.

2. If your branch does have merge commits:

   a. Pull latest `dev`, `hambach` and `redwood` branches

   b. Create new branches for backports and merge `dev` (replace `<PR#>` with your PR #)

   ```
   git checkout -b hambach-backport-<PR#> hambach
   git merge dev
   git push origin hambach-backport-<PR#>
   git checkout -b redwood-backport-<PR#> redwood
   git merge dev
   git push origin redwood-backport-<PR#>`
   ```

   c. Open new PRs in regen-web targeting `hambach` and `redwood`, respectively.

   d. If PR is a bug fix repeat the same process for `master` branch.

### How to test

1. https://deploy-preview-1611--regen-registry.netlify.app/

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**._

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] verified React components follow DRY principles
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

flagrede commented on web-registry/src/components/organisms/BuyCreditsModal/BuyCreditsModal.tsx:
```diff
@@ -130,15 +132,16 @@ const BuyCreditsModal: React.FC<React.PropsWithChildren<BuyCreditsModalProps>> =
+    const { data: allowedDenomsData } = useQuery(
+      getAllowedDenomQuery({
+        client: marketplaceClient,
+        enabled: !!marketplaceClient,
```
It's important to use enabled for queries that require a client otherwise they will run once with an undefined client.

wgwz replied:
do we have queries that would not require a client?

flagrede replied:
Yes, coingecko or metadata ones for example.

flagrede commented on web-registry/src/lib/clients/reactQueryClient.ts:
```diff
@@ -4,6 +4,8 @@ export const reactQueryClient = new QueryClient({
     queries: {
       refetchOnWindowFocus: false,
+      refetchOnMount: false,
+      staleTime: Infinity,
```
`staleTime: Infinity` means that we never consider cache stale unless we invalidate part of it ourselves.
Note this is true only for one session, if a user refreshes their browser the cache is getting reset.
Currently, the only data we want to keep up to date during a session are sell orders. For those, cache invalidation happens in a refetch callback.

wgwz replied:
> For those, cache invalidation happens in a refetch callback.

could you point out where that happens so we can see an example? thanks ahead 🙏

flagrede replied:
https://github.com/regen-network/regen-web/pull/1611/files#diff-290478b41c2ea14cca10d871fa14a00a91187ecaa2b2b31689e58b13a4cc2071R40

flagrede replied:
I also added a comment with some additional explanations.

flagrede commented on web-registry/src/lib/ecocredit/api.ts:
```diff
@@ -307,13 +313,31 @@ export const getBatchesByProjectWithSupply = async (
+      reactQueryClient,
+    });
+  } else {
+    projectData = await getProject(projectId);
```
Keeping the old path for retro-compatibility of the function. Once all paths use the react-query one we will be able to remove it.

blushi replied:
could you add a TODO?

flagrede commented on web-registry/src/lib/ecocredit/api.ts:
```diff
@@ -307,13 +313,31 @@ export const getBatchesByProjectWithSupply = async (
+          reactQueryClient,
+        });
+      } else {
+        metadata = await getMetadata(project.metadata);
```
Same here.

flagrede commented on web-registry/src/lib/normalizers/normalizeProjectsWithOrderData.ts:
```diff
@@ -0,0 +1,55 @@
+  metadatas?: any[];
+}
+
+export const normalizeProjectsWithOrderData = ({
```
This is an existing function that has been extracted from its hook.
Also, metadata fetches there were previously done here have been moved elsewhere to keep the function without async code.

flagrede commented on web-registry/src/pages/Home/Home.loader.ts:
```diff
@@ -0,0 +1,74 @@
+  queryClient: QueryClient;
+};
+
+export const homeLoader =
```
Note that loaders are completely optional, a page should work even without them.
There's just a way to trigger queries early on.
Also queries with `await` block page render while those without do not.

flagrede commented on web-registry/src/routes.tsx:
```diff
@@ -75,119 +76,100 @@ const Activity = lazy(() => import('./pages/Activity'));
+  reactQueryClient: QueryClient;
+};
+
+export const getRoutes = ({ reactQueryClient }: RouterParams): RouteObject[] =>
```
Ignore the diff below, only this line has changed and also the last one with the router declaration.

victordelval commented on web-registry/src/lib/clients/transferQueryClient.ts:
```diff
@@ -0,0 +1,9 @@
+import { QueryClientImpl } from '@regen-network/api/lib/generated/ibc/applications/transfer/v1/query';
+
+import { connect as connectToApi } from '../../ledger';
+
+export const getTransferQueryClient = async (): Promise<QueryClientImpl> => {
+  const api = await connectToApi();
+  if (!api || !api?.queryClient) return Promise.reject();
+  return new QueryClientImpl(api.queryClient);
+};
```
maybe it would be clearer to call this client `ibcQueryClient` and `getIbcQueryClient` ?

flagrede replied:
`ibcQueryClient` won't work if we start using different clients from `ibc`, `ibcTransferQueryClient` could work it's just a bit longer. If we go with this naming convention, we will have other clients like `getEcocreditMarketplaceClient`. 
If we are fine with the length we can go with this.

victordelval replied:
mm, good point... we can leave it like that for now and discuss it, but if we are going to have more clients, perhaps we are interested in having the most complete names to identify them more easily

flagrede replied:
Yes, I guess it's a trade-off between readability and import uniqueness. Personally, I would prefer the shorter version because figuring out the good import happens only once while using the shorter version helps with readability every time we open the file.

blushi replied:
agreed it's ok for now but I guess in the future, we could just use subfolders instead of long names to mimic the proto generated file structures, eg:
`clients/regen/ecocredit/queryClient.ts`
`clients/regen/ecocredit/marketplace/queryClient.ts`
`clients/ibc/applications/transfer/queryClient.ts`

victordelval commented on web-registry/src/lib/queries/react-query/coingecko/simplePrice/simplePriceQuery.ts:
```diff
@@ -0,0 +1,25 @@
+}: ReactQuerySimplePriceParams): ReactQuerySimplePriceResponse => ({
+  queryKey: ['simplePrice', GECKO_TOKEN_IDS],
+  queryFn: async () => {
+    const simplePrice = await fetchSimplePrice({
+      ids: GECKO_TOKEN_IDS,
+      vsCurrencies: GECKO_USD_CURRENCY,
+    });
```
It seems that we use this function by default with those parameters, which seems more like our default configuration... maybe it would make sense to export from `lib/coingecko` a wrapper function with the parameters already set (option 1), or even the `fetchSimplePrice` function itself with the parameters optional and a default value (option 2). In this way we could call it directly without parameters like:
```
const simplePrice = await fetchSimplePrice();
or
const simplePrice = await fetchSimplePrice({});
```
* option 1
```
export const fetchSimplePriceDefault =
  async (): Promise<FetchSimplePriceResponse> =>
    await fetchSimplePrice({
      ids: GECKO_TOKEN_IDS,
      vsCurrencies: GECKO_USD_CURRENCY,
    });
```

* option 2
```
type FetchSimplePriceParams = {
  ids?: string;
  vsCurrencies?: string;
};

export const fetchSimplePrice = async ({
  ids = GECKO_TOKEN_IDS,
  vsCurrencies = GECKO_USD_CURRENCY,
}: FetchSimplePriceParams): Promise<FetchSimplePriceResponse> => {
...
```

flagrede replied:
Option 2 might make sense I think, it still allows doing custom calls while simplifying the default case.

flagrede replied:
I've implemented option 2.

victordelval commented on web-registry/src/lib/queries/react-query/utils/getFromCacheOrFetch.ts:
```diff
@@ -0,0 +1,19 @@
+  query: QueryObserverOptions<QueryResponse>;
+}
+
+export async function getFromCacheOrFetch<QueryResponse>({
```
perhaps the name `checkCacheOrFetch` would be clearer (since nothing is returned at this point) ?

victordelval replied:
oh right, I see..

flagrede commented on web-registry/src/lib/queries/react-query/ecocredit/queries.ecocredit.ts:
```diff
@@ -12,7 +12,7 @@ import {
 export const getBatchesQuery = ({
   client,
   request,
-  enabled,
+  enabled = true,
 }: ReactQueryBatchesProps): ReactQueryBatchesResponse => ({
   queryKey: [
     'batches',
```
Yes, this file should be deleted. I started the previous PR with a single file and quickly realized it didn't scale well. Consider this file to be removed.

wgwz commented:
> - simplify queries creation

could you point out in comments which changes apply to this point you added?

flagrede commented:
> could you add some docs in the readme about how to use the react-query dev tool?

Not sure what I would write as doc. Basically, when you run registry on dev, you will have a small icon in the bottom left corner. When you click on it, it opens the devtool with all the requests that have been done with react-query.

![image](https://user-images.githubusercontent.com/530644/204860781-7e14529d-1635-4722-8746-2f2297170b39.png)

flagrede commented:
> > * simplify queries creation
> 
> could you point out in comments which changes apply to this point you added? (a single comment may be enough)

It was mainly Typescript improvements. Before we had some boilerplate because I didn't find at first how to type properly all react-query options (this is now solved). I also made a small ts utils `ReactQueryBuilderResponse` to make it easier to type the response.
Now creating a new query can be by done copying an existing query and just changing a few types in it.

flagrede commented:
> > * use only one instance of ReactQuery across the app
> 
> could you point out how we're doing this in a comment as well?

Here you go: https://github.com/regen-network/regen-web/pull/1611/files#r1036234727

flagrede commented on web-registry/src/pages/Home/Home.loader.ts:
```diff
@@ -0,0 +1,74 @@
+    });
+
+    // Optionnal data
+    getFromCacheOrFetch({
```
I left a comment about that here: https://github.com/regen-network/regen-web/pull/1611#discussion_r1031618403
You got it right, basically requests with `await` block the page to render while the ones without it can be done async.

wgwz replied:
Thanks, sorry, didn't connect the two. Makes sense.

flagrede commented on web-registry/src/lib/queries/react-query/ecocredit/getAddDataToBatchesQuery/getAddDataToBatchesQuery.types.ts:
```diff
@@ -0,0 +1,18 @@
+
+import { ReactQueryBuilderResponse } from '../../types/react-query.types';
+
+export type ReactQueryAddDataToBatchesResponse = QueryObserverOptions<
```
Using `QueryObserverOptions` here allows us to have all possible options for a query.

flagrede commented on web-registry/src/lib/queries/react-query/ecocredit/getAddDataToBatchesQuery/getAddDataToBatchesQuery.types.ts:
```diff
@@ -0,0 +1,18 @@
+> & {
+  batches?: AddDataToBatchesParams['batches'];
+  reactQueryClient?: QueryClient;
+} & ReactQueryBuilderResponse<ReactQueryAddDataToBatchesResponse>;
```
`ReactQueryBuilderResponse` is a small ts utils to simplify query creation.

wgwz replied:
i'm having some trouble understanding how this part works, but i'm curious to understand it better. maybe it's something you can cover tomorrow in the engineering check-in or some other time, definitely not a blocker

flagrede commented on web-registry/src/features/marketplace/BuySellOrderFlow/hooks/useFetchSellOrders.ts:
```diff
@@ -0,0 +1,49 @@
+  const refetchSellOrders = useCallback(async (): Promise<
+    SellOrderInfoExtented[] | undefined
+  > => {
+    await reactQueryClient.invalidateQueries({
```
Here invalidating the cache will automatically run again all query using this key.
Below we're returning the new result from the cache because this callback needs to return the new value but in most cases reload callbacks just need to invalidate a key is enough to refresh data and the UI.

blushi replied:
I think we could add that as comment in the code itself

wgwz commented on web-registry/src/pages/Home/Home.loader.ts:
```diff
@@ -0,0 +1,74 @@
+    const simplePriceQuery = getSimplePriceQuery({});
+    const projectsQuery = getProjectsQuery({
+      client: ecocreditClient,
+      request: {},
```
do you think it would be an improvement to make an empty `{}` the default value for `request` and make it optional to pass? at least here the call would be marginally simpler

flagrede replied:
It would but I'm reusing existing ones for most queries, changing that would involve a lot of changes at once.

wgwz commented on web-registry/src/pages/Home/Home.loader.ts:
```diff
@@ -0,0 +1,74 @@
+    // Mandatory data
+    await getFromCacheOrFetch({
+      query: allHomePageQuery,
+      reactQueryClient: queryClient,
```
just a thought i wanted to share, we have this same line quite a few times, and i was thnking that if we wanted to reduce that duplication, we could refactor this into class, where we move the `queryClient` into a constructor. but i don't feel very strongly about implementing that right now

flagrede replied:
That would help with the loader case indeed where there's a lot of usages maybe a bit less for a single call `(new getFromCacheOrFetch(queryClient)).getQuery(allHomePageQuery)`.
I'll give it some thought, we can always improve the loaders later on with something like that and keep the other call elsewhere.

wgwz commented:
> > could you add some docs in the readme about how to use the react-query dev tool?
> 
> Not sure what I would write as doc. Basically, when you run registry on dev, you will have a small icon in the bottom left corner. When you click on it, it opens the devtool with all the requests that have been done with react-query.
> 
> ![image](https://user-images.githubusercontent.com/530644/204860781-7e14529d-1635-4722-8746-2f2297170b39.png)

LGTM. I tried it out locally, it's pretty clear what it's for once you click it. Agreed then, no need to document.

wgwz commented on web-registry/src/pages/Home/Home.loader.ts:
```diff
@@ -0,0 +1,74 @@
+  async ({ params }: { params: any }) => {
+    // Clients
+    const ecocreditClient = await getEcocreditQueryClient();
+    const marketplaceClient = await getMarketplaceQueryClient();
```
what's different about these clients? when i'm looking at the definitions they both appear to be set up exactly the same way. i couldn't spot the difference.. maybe a comment here to explain the intention?

wgwz replied:
oh nevermind, i see it now. 🤦

wgwz commented on web-registry/src/pages/Home/Home.loader.ts:
```diff
@@ -0,0 +1,74 @@
+      client: ecocreditClient,
+      request: {},
+    });
+    const sellOrdersQuery = getSellOrdersExtentedQuery({
```
nit; `getSellOrdersExtentedQuery` => `getSellOrdersExtendedQuery`

wgwz reviewed (approved):
looks great! pre-approving. while i don't understand all the details yet, i do see the general idea and like the direction this is headed. pulled it down locally and checked out the react-query-devtools while navigating around 👍

flagrede commented:
@wgwz I think I addressed all your comments. I will proceed to merge it so I can rebase the second branch on the portfolio page. Let me know if you have additional feedback, I will be able to handle them there.

flagrede commented:
@Mergifyio backport redwood hambach

blushi reviewed (commented):
Great work!
Just a small comment: what about putting `queries/react-query/marketplace` under `queries/react-query/ecocredit` to be more consistent with the current structure under `lib/ecocredit` and mainly the one from @regen-network/api and Regen Ledger (marketplace being a "submodule" of ecocredit)
